scripts/helper_file_generator.py

```diff
@@ -1000,11 +1001,11 @@ def format_info(ext_name, info):
             struct.extend([
                 '',
                 '        static const std::vector<const char *> V_1_1_promoted_%s_apis = {' % type.lower() ])
-            struct.extend(['            %s,' % extension_dict[ext_name]['define'] for ext_name in promoted_1_1_ext_list])
+            struct.extend(['            %s_EXTENSION_NAME,' % (promoted_ext_special_defines[ext_name].upper() if ext_name in promoted_ext_special_defines.keys() else ext_name.upper()) for ext_name in promoted_1_1_ext_list])
```

Comment:
The extension_dict here is just used to grab the correct define for the promoted extensions that we're trying to print out here. We don't technically need to use it, but it provides a convenient way to convert from extension name to the extension define (this is where those issues with "maintenance1" vs "MAINTENANCE_1" come up). The problem here is that extension_dict doesn't actually contain the promoted extensions when generating source for vksc (it does when generating for vk).

My previous script edits just naively removed all the extensions in the promoted_#_#_ext_list's (see stuff removed around helper_file_generator.py:879), so the edited/red line worked previously only because there were no extensions in promoted_1_1_ext_list. None of the promoted extensions are actually in the extension_dict for vksc as it's built from extensions passed down by reg.py which filters extensions based on the 'supported' field in vk.xml, meaning that if promoted_1_1_ext_list did contain an extension, we would hit a key error here in the script as extension_dict will not contain any of the promoted extensions.

In this cl, we decided to leave in references to promoted extensions and use the vksc_compatibility.h header to work around any build issues. That meant removing the naïve code mentioned above and actually using the extensions in promoted_#_#_ext_list's. That however, creates a problem, as extension_dict will not actually contain any of these promoted extensions for vksc. Resolving that would involve edits to reg.py, which belongs to the headers repo, and I'm not sure if edits there would be cleaner.

To work around that, I check if the promoted extension is in the special table that contains conversions for extensions that have a difference in their name and define (like "maintenance1" vs "MAINTENANCE_1"). If it is, we use the define from the conversion table and otherwise we just use the normal extension name. I'm still not really sure why these differences exist in the first place.

I'm open to other ideas, but this was the best I could come up with after tinkering for awhile.

Reply:
ok.